### PR TITLE
OAK-8985: [oak-search-elastic] use bulk API for indexing

### PR DESCRIPTION
**Pull request by fabriziofortino**
* Code has been refactored/simplified (`ElasticsearchDescriptor` functions moved to `ElasticsearchIndexDefinition`)
* `ElasticsearchIndexWriter` now uses ES bulk processor
* Tests have been reworked to cope with index async nature

### CODE REVIEW

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java [hunk 1]
```diff
@@ -16,6 +16,21 @@
  */
 package org.apache.jackrabbit.oak.plugins.index.elasticsearch;
 
-public class ElasticsearchIndexConstants {
-    public static final String TYPE_ELASTICSEARCH = "elasticsearch";
+public interface ElasticsearchIndexConstants {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java]: Using interface solely for defining constants has been considered bad practice. We should consider using final classes with private constructor for defining constants.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java [hunk 2]
```diff
@@ -18,11 +18,100 @@
...
+
+    /**
+     * Convert {@code e} to Elasticsearch safe index name.
+     * Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html
+     */
+    private static String getESSafeName(String suggestedIndexName) {
+        return suggestedIndexName.replaceAll(INVALID_CHARS_REGEX, "").toLowerCase();
+    }
+
+    public static class Builder extends IndexDefinition.Builder {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:126]: There is also a class named `ElasticsearchIndexDefinitionBuilder` which doesn't seem to be used anywhere except tests. The ElasticSearch classes extend two different base classes -
1. `org.apache.jackrabbit.oak.plugins.index.search.util.IndexDefinitionBuilder`
2. `org.apache.jackrabbit.oak.plugins.index.search.IndexDefinition.Builder`

Not sure why we need two different classes. Providing some javadoc would really help :)

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java [hunk 3]
```diff
@@ -18,11 +18,100 @@
...
+        @Override
+        public ElasticsearchIndexDefinition build() {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:128]: Why do we need to change the return type here? Are we using the `build` method at a place where we need something which is not present in the base class `IndexDefinition` ? If it is for using the method `getElasticsearchIndexName`, probably we can get rid of that (see comment on that function).

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java [hunk 4]
```diff
@@ -18,11 +18,100 @@
...
+            return (ElasticsearchIndexDefinition) super.build();
+        }
+
+        @Override
+        public ElasticsearchIndexDefinition.Builder reindex() {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:133]: same question as above.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java [hunk 5]
```diff
@@ -18,11 +18,100 @@
...
+        this.elasticsearchIndexName = getRemoteIndexName();
+
+        this.bulkActions = getOptionalValue(defn, BULK_ACTIONS, BULK_ACTIONS_DEFAULT);
+        this.bulkSizeBytes = getOptionalValue(defn, BULK_SIZE_BYTES, BULK_SIZE_BYTES_DEFAULT);
+        this.bulkFlushIntervalMs = getOptionalValue(defn, BULK_FLUSH_INTERVAL_MS, BULK_FLUSH_INTERVAL_MS_DEFAULT);
+        this.bulkRetries = getOptionalValue(defn, BULK_RETRIES, BULK_RETRIES_DEFAULT);
+        this.bulkRetriesBackoff = getOptionalValue(defn, BULK_RETRIES_BACKOFF, BULK_RETRIES_BACKOFF_DEFAULT);
+    }
+
+    public String getElasticsearchIndexName() {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java]: Can we override the base class method `getIndexName` instead of creating a new method?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/query/ElasticsearchIndex.java [hunk 6]
```diff
@@ -48,7 +48,7 @@
     private final ElasticsearchConnection elasticsearchConnection;
     private final NodeState root;
 
-    ElasticsearchIndex(@NotNull ElasticsearchConnection elasticsearchConnection, @NotNull NodeState root) {
+    protected ElasticsearchIndex(@NotNull ElasticsearchConnection elasticsearchConnection, @NotNull NodeState root) {
```
Comment (averma21) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/query/ElasticsearchIndex.java]: This doesn't seem to be used outside package. Can we revert the change to access modifier.

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java [hunk 7]
```diff
@@ -118,12 +117,15 @@ public void indexSelection() throws Exception {
...
-        assertQuery("select [jcr:path] from [nt:base] where [propa] = 'foo2'", singletonList("/test/c"));
-        assertQuery("select [jcr:path] from [nt:base] where [propc] = 'foo'", singletonList("/test/d"));
+        assertEventually(() -> {
+            assertThat(explain(propaQuery), containsString("elasticsearch:test1"));
+            assertThat(explain("select [jcr:path] from [nt:base] where [propc] = 'foo'"), containsString("elasticsearch:test2"));
+
+            assertQuery(propaQuery, Arrays.asList("/test/a", "/test/b"));
+            assertQuery("select [jcr:path] from [nt:base] where [propa] = 'foo2'", singletonList("/test/c"));
+            assertQuery("select [jcr:path] from [nt:base] where [propc] = 'foo'", singletonList("/test/d"));
+        }, 10000);
```
Comment (averma21) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java]: Should we define a constant for `10000`? It seems to be used at multiple places. If this is related to BULK_FLUSH_INTERVAL_MS_DEFAULT, may be we could use an expression to calculate it like -
```
private static final int TIMEOUT  = BULK_FLUSH_INTERVAL_MS_DEFAULT * 3;
```

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java (hunk 1 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java]: That's correct. I have removed `ElasticsearchIndexConstants ` at all.
I moved the constants in the actual class where they are most relevant (`ElasticsearchIndexDefinition`). Having a 'bag of constants' class/interface isn't really following OO best practices neither. For example, in the Java API, every class/interface has its constants. They are not in a Constants class/interface with hundreds of constants.

Every strategy has pros/cons on this topic. I think this highlights a lack of the language itself. Other languages introduced different constructs to overcome this in a more elegant way. An example is `companion objects`.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java [hunk 8]
```diff
@@ -1,21 +0,0 @@
...
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package org.apache.jackrabbit.oak.plugins.index.elasticsearch;
-
-public class ElasticsearchIndexConstants {
-    public static final String TYPE_ELASTICSEARCH = "elasticsearch";
```
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java:20]: It might make sense to keep this file for constants (since these would get used a lot in tests even outside the elastic module) - I think Using ElasticsearchIndexConstants.TYPE_ELASTICSEARCH sounds better than using ElasticsearchIndexDefinition.TYPE_ELASTICSEARCH

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 5 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java]: there is no 1 to 1 mapping between the OAK index name and index name in ES. In the future, for scalability reasons, we could also have multiple indexes in ES related to an OAK index definition. We cannot override `getIndexName`. I will add some Javadoc on `getElasticsearchIndexName`. Maybe a different name would help? Something like `getRemoteIndexName`?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java (hunk 8 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexConstants.java:20]: Please see my comment here https://github.com/oak-indexing/jackrabbit-oak/pull/94#discussion_r400836707

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/query/ElasticsearchIndex.java (hunk 6 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/query/ElasticsearchIndex.java]: Correct. I reverted it and changed most of the class modifiers as well. Most of them, in fact, should be `package private`. I came to this conclusion looking at Java API code and watching this [video](https://www.youtube.com/watch?v=KrLFs6f2bOA) (the first 20 min).

There are a lot of advantages with this strategy. One of the features, that I find quite nice, is the following: in one look you understand what a single package exposes:
Screenshot 2020-03-31 at 16 01 17

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 3 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:128]: Yes, we need to change the return type because we need to access elastic specific methods.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 4 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:133]: Yes, we need to change the return type because we need to access elastic specific methods.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 2 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:126]: This is something we have inherited from the initial prototype that it's not completely clear to me. Despite the similarity in the names, these classes are needed for different reasons. I know the one in the `util` package is only needed for tests now but my understanding is that this could be useful when I have to create an elasticsearch index definition from outside (test, benchmark, etc).

I have seen different ways to create an index definition. Not sure what is the more appropriate though.

@nit0906 @thomasmueller any advice on that? I would like to make it right now that we don't have many dependencies on this class.

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 5 above)
Comment (averma21) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java]: Yes. `getRemoteIndexName` would clear the misunderstanding :)

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java [hunk 9]
```diff
@@ -18,11 +18,122 @@
...
+
 public class ElasticsearchIndexDefinition extends IndexDefinition {
+
+    public static final String TYPE_ELASTICSEARCH = "elasticsearch";
+
+    public static final String BULK_ACTIONS = "bulkActions";
+    public static final int BULK_ACTIONS_DEFAULT = 250;
+
+    public static final String BULK_SIZE_BYTES = "bulkSizeBytes";
+    public static final long BULK_SIZE_BYTES_DEFAULT = 2_097_152; // 2MB
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java]: What about 2 * 1024 * 1024?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java (hunk 2 above)
Comment (thomasmueller) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:126]: From how it's used, it looks like IndexDefinitionBuilder is used to create the index definition node structure (but not during indexing), and IndexDefinition.Builder is used to create an IndexDefinition object (but not the node structure). Adding javadocs would help!

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java [hunk 10]
```diff
@@ -204,4 +213,32 @@ private String explain(String query) {
         String explain = "explain " + query;
         return executeQuery(explain, "JCR-SQL2").get(0);
     }
+
+    private static void assertEventually(Runnable r, long timeoutMillis) {
```
Comment (thomasmueller) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java:221]: Good! Adding a convenience method with just one parameter (Runnable) would be nice, because all uses currently use BULK_FLUSH_INTERVAL_MS_DEFAULT * 3 AFAIK.
